gslb/k8sobjects/multicluster_ingress_objects.go

```diff
@@ -0,0 +1,351 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*   http://www.apache.org/licenses/LICENSE-2.0
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package k8sobjects
+
+import (
+	"errors"
+	"sort"
+	"sync"
+
+	"github.com/vmware/global-load-balancing-services-for-kubernetes/gslb/gslbutils"
+	akov1alpha1 "github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/apis/ako/v1alpha1"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
+)
+
+var mcihMapInit sync.Once
+var mcihMap ObjHostMap
+
+func getMultiClusterIngressHostMap() *ObjHostMap {
+	mcihMapInit.Do(func() {
+		mcihMap.HostMap = make(map[string]IPHostname)
+	})
+	return &mcihMap
+}
+
+// GetHostMetaForMultiClusterIngress returns a multi-cluster ingress split into its backends
+func GetHostMetaForMultiClusterIngress(mci *akov1alpha1.MultiClusterIngress, cname string) []MultiClusterIngressHostMeta {
+	metaObjects := []MultiClusterIngressHostMeta{}
+	// hostIPList := gslbutils.IngressGetIPAddrs(ingress)
+	// TODO: move the below logic to utils
+	var hostIPList []gslbutils.IngressHostIP
+	for _, ing := range mci.Status.LoadBalancer.Ingress {
+		ingHostIP := gslbutils.IngressHostIP{
+			Hostname: ing.Hostname,
+			IPAddr:   ing.IP,
+		}
+		hostIPList = append(hostIPList, ingHostIP)
+	}
+
+	tlsHosts := getMultiClusterIngressTLSHosts(mci)
+
+	gf := gslbutils.GetGlobalFilter()
+
+	// we don't return because of errors here, as we need these objects in the our internal cache,
+	// so that, when the GDP object gets changed, we can re-apply these objects back again.
+	// The errors for syncVIPsOnly are taken care of in the graph layer.
+	syncVIPsOnly, err := gf.IsClusterSyncVIPOnly(cname)
+	if err != nil {
+		gslbutils.Logf("cluster: %s, ns: %s, ingress: %s, msg: skipping ingress because of error: %v",
+			cname, mci.Namespace, mci.Name, err)
+	}
+	var vsUUIDs map[string]string
+	var controllerUUID string
+
+	vsUUIDs, controllerUUID, err = parseVSAndControllerAnnotations(mci.ObjectMeta.Annotations)
+	if err != nil && !syncVIPsOnly {
+		// Note that the ingress key will still be published to graph layer, but the key
+		// won't be processed, this is just to maintain the ingress information as part
+		// of in-memory map.
+		gslbutils.Errf("ns: %s, ingress: %s, msg: skipping ingress because of error: %v",
+			mci.Namespace, mci.Name, err)
+	}
+	if (controllerUUID == "" || len(vsUUIDs) == 0) && !syncVIPsOnly {
+		gslbutils.Logf("cluster: %s, ns: %s, ingress: %s, msg: skipping ingress because controller UUID absent in annotations",
+			cname, mci.Namespace, mci.Name)
+	}
+	for _, hip := range hostIPList {
```

Comment:
I think one MCI object can only have one host. In that case hostIPList would have only one value. Probably we don;t need a list here. Though its fine this way too.